File: GroundZero/ground_zero_flutter/ground_zero_core/lib/models/paged_result.dart
class PagedResult<T> {
  final List<T> items;
  final int totalCount;
  final int page;
  final int pageSize;
  final int totalPages;
  final bool hasPreviousPage;
  final bool hasNextPage;

  const PagedResult({required this.items, required this.totalCount, required this.page,
    required this.pageSize, required this.totalPages, required this.hasPreviousPage, required this.hasNextPage});

  factory PagedResult.fromJson(Map<String, dynamic> json, T Function(Map<String, dynamic>) fromJsonT) => PagedResult(
    items: (json['items'] as List).map((e) => fromJsonT(e as Map<String, dynamic>)).toList(),
    totalCount: json['totalCount'] ?? 0, page: json['page'] ?? 1, pageSize: json['pageSize'] ?? 10,
    totalPages: json['totalPages'] ?? 1, hasPreviousPage: json['hasPreviousPage'] ?? false, hasNextPage: json['hasNextPage'] ?? false);
}

File: GroundZero/GroundZero/GroundZero.Application/Common/Models/PagedResult.cs
namespace GroundZero.Application.Common.Models;

public class PagedResult<T>
{
    public List<T> Items { get; set; } = [];
    public int TotalCount { get; set; }
    public int Page { get; set; }
    public int PageSize { get; set; }
    public int TotalPages => (int)Math.Ceiling(TotalCount / (double)PageSize);
    public bool HasPreviousPage => Page > 1;
    public bool HasNextPage => Page < TotalPages;
    public PagedResult() { }
    public PagedResult(List<T> items, int totalCount, int page, int pageSize)
    { Items = items; TotalCount = totalCount; Page = page; PageSize = pageSize; }
    public PagedResult<TDest> Map<TDest>(Func<T, TDest> selector) =>
        new(Items.Select(selector).ToList(), TotalCount, Page, PageSize);
}